### PR TITLE
feat(ops): Fast zero copy string arguments

### PR DESCRIPTION
**Pull request by littledivy**
Uses SeqOneByteString optimization to do zero-copy `&str` arguments in fast calls.

- [x] Depends on https://github.com/denoland/rusty_v8/pull/1129
  - [x] Depends on https://chromium-review.googlesource.com/c/v8/v8/+/4036884
- [x] Disable in async ops
- [x] Make it work with owned `String` with an extra alloc in fast path.
- [x] Support `Cow<'_, str>`. Owned for slow case, Borrowed for fast case 

```rust
#[op]
fn op_string_len(s: &str) -> u32 { 
  str.len() as u32 
}
```

### CODE REVIEW

**Comment by bartlomieju** [comment]
> I recommend testing some Latin-1 (V8's OneByte) and UTF-8 clashing characters, eg. © ("\u00A9") is a good one.
> 
> EDIT: Not sure if that's actually a clashing character... But at least I've recently been bitten by presuming that V8's OneByteStrings are directly usable as UTF-8, specifically with the copyright symbol as the test character that broke.

@littledivy did you add such a test?

File: cli/bench/console.js [hunk 1]
```diff
@@ -1,8 +1,3 @@
 // Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
 const count = 100000;
-
-const start = Date.now();
 for (let i = 0; i < count; i++) console.log("Hello World");
-const elapsed = Date.now() - start;
-const rate = Math.floor(count / (elapsed / 1000));
-console.log(`time ${elapsed} ms rate ${rate}`);
```
Comment (ry) [inline cli/bench/console.js:8]: just using time? why this change?

File: ext/web/lib.rs [hunk 2]
```diff
@@ -352,25 +352,42 @@ impl Resource for TextDecoderResource {
...
+    boundary
+  };
+
+  buffer[..boundary].copy_from_slice(input[..boundary].as_bytes());
+
+  // The `read` output parameter is measured in UTF-16 code units.
+  out_buf[0] = match input {
+    // ops guarantee this to be a SeqOneByteString.
+    Cow::Borrowed(v) => v[..boundary].len() as u32,
+    Cow::Owned(v) => v[..boundary].encode_utf16().count() as u32,
```
Comment (ry) [inline ext/web/lib.rs:388]: `Cow::Borrowed` means `SeqOneByteString` and `Cow::Owned` means utf16? I think a more verbose comment here is warranted.

File: ext/web/lib.rs [hunk 3]
```diff
@@ -352,25 +352,42 @@ impl Resource for TextDecoderResource {
...
+  };
+  out_buf[1] = boundary as u32;
```
Comment (ry) [inline ext/web/lib.rs:390]: Do we have sufficient unit tests to cover all these cases?

File: ops/lib.rs [hunk 4]
```diff
@@ -618,14 +635,25 @@ fn is_result(ty: impl ToTokens) -> bool {
...
 }
 
 fn is_option_string(ty: impl ToTokens) -> bool {
   tokens(ty) == "Option < String >"
 }
 
+fn is_cow_str(ty: impl ToTokens) -> bool {
+  tokens(&ty).starts_with("Cow <") && tokens(&ty).ends_with("str >")
+}
+
```
Comment (ry) [inline ops/lib.rs:656]: `is_string` shouldn't also include Cow strings? Maybe it should be called `is_non_cow_string`?

It would be nice to have unit tests for these little utility functions. Is that hard?

**Comment by littledivy** [comment]
@bartlomieju There are tests for that in the V8 patch: https://chromium-review.googlesource.com/c/v8/v8/+/4036884

File: cli/bench/console.js (hunk 1 above)
Comment (littledivy) [reply cli/bench/console.js:8]: Removing extra stuff so it can easily measured using `time` and `hyperfine`.

**Review by ry — APPROVED** [review]
LGTM - some comments...

File: ext/web/lib.rs (hunk 3 above)
Comment (littledivy) [reply ext/web/lib.rs:390]: we've got WPT for TextEncoder

File: ext/web/lib.rs (hunk 2 above)
Comment (bartlomieju) [reply ext/web/lib.rs:388]: +1 please explain which variant should be used in which situation
